release/src/router/lldpd-0.9.8/src/client/lldpcli.8.in
.\" Copyright (c) 2006 Pierre-Yves Ritschard <pyr@openbsd.org>
.\" Copyright (c) 2008 Vincent Bernat <bernat@luffy.cx>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: July 16 2008 $
.Dt LLDPCLI 8
.Os
.Sh NAME
.Nm lldpcli ,
.Nm lldpctl
.Nd control LLDP daemon
.Sh SYNOPSIS
.Nm
.Op Fl dv
.Op Fl u Ar socket
.Op Fl f Ar format
.Op Fl c Ar file
.Op Ar command ...
.Nm lldpctl
.Op Fl dv
.Op Fl u Ar socket
.Op Fl f Ar format
.Op Ar interfaces ...
.Sh DESCRIPTION
The
.Nm
program controls
.Xr lldpd 8
daemon.
.Pp
When no command is specified,
.Nm
will start an interactive shell which can be used to input arbitrary
commands as if they were specified on the command line. This
interactive shell should provide completion and history support.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Enable more debugging information. This flag can be repeated.
.It Fl u Ar socket
Specify the Unix-domain socket used for communication with
.Xr lldpd 8 .
.It Fl v
Show
.Nm
version. When repeated, show more build information.
.It Fl f Ar format
Choose the output format. Currently
.Em plain ,
.Em xml ,
.Em json ,
.Em json0
and
.Em keyvalue
formats are available. The default is
.Em plain .
.Em json0
is more verbose than
.Em json
but the structure of the JSON object is not affected by the number of
interfaces or the number of neighbors. It is therefore easier to
parse.
.It Fl c Ar file
Read the given configuration file. This option may be repeated several
times. If a directory is provided, each file contained in it will be
read  if ending by
.Li .conf .
Order is alphabetical.
.El
.Pp
When invoked as
.Nm lldpctl ,
.Nm
will display detailed information about each neighbors on the
specified interfaces or on all interfaces if none are specified. This
command is mostly kept for backward compatibility with older versions.
.Pp
The following commands are supported by
.Nm .
When there is no ambiguity, the keywords can be abbreviated. For
example,
.Cd show neighbors ports eth0 summary
and
.Cd sh neigh p eth0 sum
are the same command.
.Bd -ragged -offset XX
.Cd exit
.Bd -ragged -offset XXXXXX
Quit
.Nm .
.Ed

.Cd help Op ...
.Bd -ragged -offset XXXXXX
Display general help or help about a command. Also, you can get help
using the completion or by pressing the
.Ic ?
key. However, completion and inline help may be unavailable if
.Nm
was compiled without readline support but
.Cd help
command is always available.
.Ed

.Cd show neighbors
.Op ports Ar ethX Op ,...
.Op Cd details | summary
.Op Cd hidden
.Bd -ragged -offset XXXXXX
Display information about each neighbor known by
.Xr lldpd 8
daemon. With
.Cd summary ,
only a the name and the port description of each remote host will be
displayed. On the other hand, with
.Cd details ,
all available information will be displayed, giving a verbose
view. When using
.Cd hidden ,
also display remote ports hidden by the smart filter. When specifying
one or several ports, the information displayed is limited to the
given list of ports.
.Ed

.Cd show chassis
.Op Cd details | summary
.Bd -ragged -offset XXXXXX
Display information about local chassis. With
.Cd summary ,
most details are skipped. On the other hand, with
.Cd details ,
all available information will be displayed, giving a verbose
view.
.Ed

.Cd watch
.Op ports Ar ethX Op ,...
.Op Cd details | summary
.Op Cd hidden
.Op Cd limit Ar X
.Bd -ragged -offset XXXXXX
Watch for any neighbor changes and report them as soon as they
happen. When specifying ports, the changes are only reported when
happening on the given ports.
.Cd hidden , summary
and
.Cd details
have the same meaning than previously described. If
.Cd limit
is specificed,
.Nm
will exit after receiving the specified number of events.
.Ed

.Cd show configuration
.Bd -ragged -offset XXXXXX
Display global configuration of
.Xr lldpd 8
daemon.
.Ed

.Cd show statistics
.Op ports Ar ethX Op ,...
.Op Cd summary
.Bd -ragged -offset XXXXXX
Report LLDP-related statistics, like the number of LLDPDU transmitted,
received, discarded or unrecognized. When specifying ports, only the
statistics from the given port are reported. With
.Cd summary
the statistics of each port is summed.
.Ed

.Cd update
.Bd -ragged -offset XXXXXX
Make
.Xr lldpd 8
update its information and send new LLDP PDU on all interfaces.
.Ed

.Cd configure
.Cd system hostname Ar name
.Bd -ragged -offset XXXXXX
Override system hostname with the provided value. By default, the
system name is the FQDN found from the resolved value of
.Ic uname -n .
As a special value, use "." (dot) to use the short hostname instead of
a FQDN.
.Ed

.Cd unconfigure
.Cd system hostname
.Bd -ragged -offset XXXXXX
Do not override system hostname and restore the use of the node name.
.Ed

.Cd configure
.Cd system description Ar description
.Bd -ragged -offset XXXXXX
Override chassis description with the provided value instead of using
kernel name, node name, kernel version, build date and architecture.
.Ed

.Cd unconfigure
.Cd system description
.Bd -ragged -offset XXXXXX
Do not override chassis description and use a value computed from node
name, kernel name, kernel version, build date and architecture instead.
.Ed

.Cd configure
.Cd system platform Ar description
.Bd -ragged -offset XXXXXX
Override platform description with the provided value instead of using
kernel name. This value is currently only used for CDP.
.Ed

.Cd unconfigure
.Cd system platform
.Bd -ragged -offset XXXXXX
Do not override platform description and use the kernel name. This
option undoes the previous one.
.Ed

.Cd configure
.Cd system interface pattern Ar pattern
.Bd -ragged -offset XXXXXX
Specify which interface to listen and send LLDPDU to. Without this
option,
.Nm lldpd
will use all available physical interfaces. This option can use
wildcards. Several interfaces can be specified separated by commas.
It is also possible to blacklist an interface by suffixing it with an
exclamation mark. It is possible to whitelist an interface by
suffixing it with two exclamation marks. A whitelisted interface beats
a blacklisted interfaces which beats a simple matched interface. For
example, with
.Em eth*,!eth1,!eth2
.Nm lldpd
will only use interfaces starting by
.Em eth
with the exception of
.Em eth1
and
.Em eth2 .
While with
.Em *,!eth*,!!eth1
.Nm
will use all interfaces, except interfaces starting by
.Em eth
with the exception of
.Em eth1 .
When an exact match is found, it will circumvent some tests. For example, if
.Em eth0.12
is specified, it will be accepted even if this is a VLAN interface.
.Ed

.Cd unconfigure
.Cd system interface pattern
.Bd -ragged -offset XXXXXX
Remove any previously configured interface pattern and use all
physical interafces. This option undoes the previous one.
.Ed

.Cd configure
.Cd system interface description
.Bd -ragged -offset XXXXXX
Some OS allows the user to set a description for an interface. Setting
this option will enable
.Nm lldpd
to override this description with the name of the peer neighbor if one
is found or with the number of neighbors found.
.Ed

.Cd unconfigure
.Cd system interface description
.Bd -ragged -offset XXXXXX
Do not update interface description with the name of the peer
neighbor. This option undoes the previous one.
.Ed

.Cd configure
.Cd system interface promiscuous
.Bd -ragged -offset XXXXXX
Enable promiscuous mode on managed interfaces.
.Pp
When the interface is not managed any more (or when quitting
.Nm lldpd ) ,
the interface is left in promiscuous mode as it is difficult to know
if someone else also put the interface in promiscuous mode.
.Pp
This option is known to be useful when the remote switch is a Cisco
2960 and the local network card features VLAN hardware
acceleration. In this case, you may not receive LLDP frames from the
remote switch. The most plausible explanation for this is the frame is
tagged with some VLAN (usually VLAN 1) and your network card is
filtering VLAN. This is not the only available solution to work-around
this problem. If you are concerned about performance issues, you can
also tag the VLAN 1 on each interface instead.
.Pp
Currently, this option has no effect on anything else than Linux. On
other OS, either disable VLAN acceleration, tag VLAN 1 or enable
promiscuous mode manually on the interface.
.Ed

.Cd unconfigure
.Cd system interface promiscuous
.Bd -ragged -offset XXXXXX
Do not set promiscuous mode on managed interfaces. This option does
not disable promiscuous mode on interfaces already using this mode.
.Ed

.Cd configure
.Cd system ip management pattern Ar pattern
.Bd -ragged -offset XXXXXX
Specify the management addresses of this system. As for interfaces
(described above), this option can use wildcards and inversions.
Without this option, the first IPv4 and the first IPv6 are used. If an
exact IP address is provided, it is used as a management address
without any check. If only negative patterns are provided, only one
IPv4 and one IPv6 addresses are chosen. Otherwise, many of them can be
selected. If you want to blacklist IPv6 addresses, you can use
.Em !*:* .
.Ed

.Cd unconfigure
.Cd system ip management pattern
.Bd -ragged -offset XXXXXX
Unset any specific pattern for matching management addresses. This
option undoes the previous one.
.Ed

.Cd configure
.Cd system bond-slave-src-mac-type Ar value
.Bd -ragged -offset XXXXXX
Set the type of src mac in lldp frames sent on bond slaves

Valid types are:
.Bl -tag -width "XXX." -compact -offset XX
.It Sy real
Slave real mac
.It Sy zero
All zero mac
.It Sy fixed
An arbitrary fixed value
.Li ( 00:60:08:69:97:ef )
.It Sy local
Real mac with locally administered bit set. If the real mac already
has the locally administered bit set, fallback to the fixed value.
.El
.Pp
Default value for
.Nm bond-slave-src-mac-type
is
.Nm local .
Some switches may complain when using one of the two other possible
values (either because
.Li 00:00:00:00:00:00
is not a valid MAC or because the MAC address is flapping from one
port to another). Using
.Sy local
might lead to a duplicate MAC address on the network (but this is
quite unlikely).
.Ed

.Cd configure
.Cd lldp agent-type
.Cd nearest-bridge | nearest-non-tpmr-bridge | nearest-customer-bridge
.Bd -ragged -offset XXXXXX
The destination MAC address used to send LLDPDU allows an agent to
control the propagation of LLDPDUs. By default, the
.Li 01:80:c2:00:00:0e
MAC address is used and limit the propagation of the LLDPDU to the
nearest bridge
.Cd ( nearest-bridge ) .
To instruct
.Nm lldpd
to use the
.Li 01:80:c2:00:00:03
MAC address instead, use
.Cd nearest-nontpmr-bridge
instead.
To use the
.Li 01:80:c2:00:00:00
MAC address instead, use
.Cd nearest-customer-bridge
instead.
.Ed

.Cd configure
.Cd lldp portidsubtype
.Cd ifname | macaddress
.Pp
.Cd configure
.Op ports Ar ethX Op ,...
.Cd lldp portidsubtype
.Cd local Ar value Op Cd description Ar description
.Bd -ragged -offset XXXXXX
Force port ID subtype. By default,
.Nm lldpd
will use the MAC address as port identifier and the interface name as
port description, unless the interface has an alias. In this case, the
interface name will be used as port identifier and the description
will be the interface alias. With this command, you can force the port
identifier to be the interface name (with
.Cd ifname ) ,
the MAC address (with
.Cd macaddress )
or a local value (with
.Cd value ) .
In the latest case, the local value should be provided.
Optionally, a port description can also be provided after the local
value.
.Ed

.Cd configure
.Cd lldp tx-interval Ar interval
.Bd -ragged -offset XXXXXX
Change transmit delay to the specified value in seconds. The transmit
delay is the delay between two transmissions of LLDP PDU. The default
value is 30 seconds.
.Ed

.Cd configure
.Cd lldp tx-hold Ar hold
.Bd -ragged -offset XXXXXX
Change transmit hold value to the specified value. This value is used
to compute the TTL of transmitted packets which is the product of this
value and of the transmit delay. The default value is 4 and therefore
the default TTL is 120 seconds.
.Ed

.Cd configure
.Op ports Ar ethX Op ,...
.Cd lldp
.Cd status Ar rx-and-tx | rx-only | tx-only | disabled
.Bd -ragged -offset XXXXXX
Configure the administrative status of the given port. By default, all
ports are configured to be in
.Ar rx-and-tx
mode. This means they can receive and transmit LLDP frames (as well as
other protocols if needed). In
.Ar rx-only
mode, they won't emit any frames and in
.Ar tx-only
mode, they won't receive any frames. In
.Ar disabled
mode, no frame will be sent and any incoming frame will be
discarded. This settings do not override the operational mode of the
main daemon. If it is configured in receive-only mode (with the
.Fl r
flag), setting any transmit mode won't have any effect.
.Ed

.Cd configure
.Cd lldp custom-tlv
.Op Cd add | replace
.Cd oui Ar oui
.Cd subtype Ar subtype
.Op Cd oui-info Ar content
.Bd -ragged -offset XXXXXX
Emit a custom TLV for OUI
.Ar oui ,
with subtype
.Ar subtype
and optionally with the bytes specified in
.Ar content .
Both
.Ar oui
and
.Ar content
should be a comma-separated list of bytes in hex format.
.Ar oui
must be exactly 3-byte long.
If
.Ar add
is specified then the TLV will be added. This is the default action.
If
.Ar replace
is specified then all TLVs with the same
.Ar oui
and
.Ar subtype
will be replaced.

.Ed

.Cd unconfigure
.Cd lldp custom-tlv
.Op Cd oui Ar oui
.Op Cd subtype Ar subtype
.Bd -ragged -offset XXXXXX
When no oui is specified, remove all previously configured custom TLV.
When OUI
.Ar oui
and subtype
.Ar subtype
is specified, remove specific instances of custom TLV.
.Ed

.Cd configure med fast-start
.Cd enable | tx-interval Ar interval
.Bd -ragged -offset XXXXXX
Configure LLDP-MED fast start mechanism. When a new LLDP-MED-enabled
neighbor is detected, fast start allows
.Nm lldpd
to shorten the interval between two LLDPDU.
.Cd enable
should enable LLDP-MED fast start while
.Cd tx-interval
specifies the interval between two LLDPDU in seconds. The default
interval is 1 second. Once 4 LLDPDU have been sent, the fast start
mechanism is disabled until a new neighbor is detected.
.Ed

.Cd unconfigure med fast-start
.Bd -ragged -offset XXXXXX
Disable LLDP-MED fast start mechanism.
.Ed

.Cd configure
.Op ports Ar ethX Op ,...
.Cd med location coordinate
.Cd latitude Ar latitude
.Cd longitude Ar longitude
.Cd altitude Ar altitude Ar unit
.Cd datum Ar datum
.Bd -ragged -offset XXXXXX
Advertise a coordinate based location on the given ports (or on all
ports if no port is specified). The format of
.Ar latitude
is a decimal floating point number followed either by
.Em N
or
.Em S .
The format of
.Ar longitude
is a decimal floating point number followed either by 
.Em E
or
.Em W .
.Ar altitude
is a decimal floating point number followed either by
.Em m
when expressed in meters or
.Em f
when expressed in floors. A space is expected between the floating
point number and the unit.
.Ar datum
is one of those values:
.Bl -bullet -compact -offset XXXXXXXX
.It
WGS84
.It
NAD83
.It
NAD83/MLLW
.El
.Pp
A valid use of this command is:
.D1 configure ports eth0 med location coordinate latitude 48.85667N longitude 2.2014E altitude 117.47 m datum WGS84
.Ed

.Cd configure
.Op ports Ar ethX Op ,...
.Cd med location address
.Cd country Ar country
.Cd Op Ar type value Op ...
.Bd -ragged -offset XXXXXX
Advertise a civic address on the given ports (or on all ports if no
port is specified).
.Ar country
is the two-letter code representing the country. The remaining
arguments should be paired to form the address. The first member of
each pair indicates the type of the second member which is a free-form
text. Here is the list of valid types:
.Bl -bullet -compact -offset XXXXXXXX
.It
language
.It
country-subdivision
.It
county
.It
city
.It
city-division
.It
block
.It
street
.It
direction
.It
trailing-street-suffix
.It
street-suffix
.It
number
.It
number-suffix
.It
landmark
.It
additional
.It
name
.It
zip
.It
building
.It
unit
.It
floor
.It
room
.It
place-type
.It
script
.El
.Pp
A valid use of this command is:
.D1 configure ports eth1 med location address country US street Qo Commercial Road Qc city Qo Roseville Qc
.Ed

.Cd configure
.Op ports Ar ethX Op ,...
.Cd med location elin
.Ar number
.Bd -ragged -offset XXXXXX

release/src/router/lldpd-0.9.8/src/client/lldpcli.c
/* -*- mode: c; c-file-style: "openbsd" -*- */
/*
 * Copyright (c) 2008 Vincent Bernat <bernat@luffy.cx>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <arpa/inet.h>
#include <libgen.h>
#include <dirent.h>
#include <signal.h>
#include <sys/queue.h>

#include "client.h"

#ifdef HAVE___PROGNAME
extern const char	*__progname;
#else
# define __progname "lldpcli"
#endif

/* Global for completion */
static struct cmd_node *root = NULL;
const char *ctlname = NULL;

static int
is_lldpctl(const char *name)
{
	static int last_result = -1;
	if (last_result == -1 && name) {
		char *basec = strdup(name);
		if (!basec) return 0;
		char *bname = basename(basec);
		last_result = (!strcmp(bname, "lldpctl"));
		free(basec);
	}
	return (last_result == -1)?0:last_result;
}

static void
usage()
{
	fprintf(stderr, "Usage:   %s [OPTIONS ...] [COMMAND ...]\n", __progname);
	fprintf(stderr, "Version: %s\n", PACKAGE_STRING);

	fprintf(stderr, "\n");

	fprintf(stderr, "-d          Enable more debugging information.\n");
	fprintf(stderr, "-u socket   Specify the Unix-domain socket used for communication with lldpd(8).\n");
	fprintf(stderr, "-f format   Choose output format (plain, keyvalue, json"
#if defined USE_XML
	    ", xml"
#endif
	    ").\n");
	if (!is_lldpctl(NULL))
		fprintf(stderr, "-c conf     Read the provided configuration file.\n");

	fprintf(stderr, "\n");

	fprintf(stderr, "see manual page lldpcli(8) for more information\n");
	exit(1);
}

static int
is_privileged()
{
	/* Check we can access the control socket with read/write
	 * privileges. The `access()` function uses the real UID and real GID,
	 * therefore we don't have to mangle with our identity. */
	return (ctlname && access(ctlname, R_OK|W_OK) == 0);
}

static char*
prompt()
{
#define CESC "\033"
	int privileged = is_privileged();
	if (isatty(STDIN_FILENO)) {
		if (privileged)
			return "[lldpcli] # ";
		return "[lldpcli] $ ";
	}
	return "";
}

static int must_exit = 0;
/**
 * Exit the interpreter.
 */
static int
cmd_exit(struct lldpctl_conn_t *conn, struct writer *w,
    struct cmd_env *env, void *arg)
{
	log_info("lldpctl", "quit lldpcli");
	must_exit = 1;
	return 1;
}

/**
 * Send an "update" request.
 */
static int
cmd_update(struct lldpctl_conn_t *conn, struct writer *w,
    struct cmd_env *env, void *arg)
{
	log_info("lldpctl", "ask for global update");

	lldpctl_atom_t *config = lldpctl_get_configuration(conn);
	if (config == NULL) {
		log_warnx("lldpctl", "unable to get configuration from lldpd. %s",
		    lldpctl_last_strerror(conn));
		return 0;
	}
	if (lldpctl_atom_set_int(config,
		lldpctl_k_config_tx_interval, -1) == NULL) {
		log_warnx("lldpctl", "unable to ask lldpd for immediate retransmission. %s",
		    lldpctl_last_strerror(conn));
		lldpctl_atom_dec_ref(config);
		return 0;
	}
	log_info("lldpctl", "immediate retransmission requested successfully");
	lldpctl_atom_dec_ref(config);
	return 1;
}

/**
 * Pause or resume execution of lldpd.
 *
 * @param conn    The connection to lldpd.
 * @param pause   1 if we want to pause lldpd, 0 otherwise
 * @return 1 on success, 0 on error
 */
static int
cmd_pause_resume(lldpctl_conn_t *conn, int pause)
{
	lldpctl_atom_t *config = lldpctl_get_configuration(conn);
	if (config == NULL) {
		log_warnx("lldpctl", "unable to get configuration from lldpd. %s",
		    lldpctl_last_strerror(conn));
		return 0;
	}
	if (lldpctl_atom_get_int(config, lldpctl_k_config_paused) == pause) {
		log_debug("lldpctl", "lldpd is already %s",
		    pause?"paused":"resumed");
		lldpctl_atom_dec_ref(config);
		return 1;
	}
	if (lldpctl_atom_set_int(config,
		lldpctl_k_config_paused, pause) == NULL) {
		log_warnx("lldpctl", "unable to ask lldpd to %s operations. %s",
		    pause?"pause":"resume",
		    lldpctl_last_strerror(conn));
		lldpctl_atom_dec_ref(config);
		return 0;
	}
	log_info("lldpctl", "lldpd should %s operations",
	    pause?"pause":"resume");
	lldpctl_atom_dec_ref(config);
	return 1;
}
static int
cmd_pause(struct lldpctl_conn_t *conn, struct writer *w,
    struct cmd_env *env, void *arg) {
	(void)w; (void)env;
	return cmd_pause_resume(conn, 1);
}
static int
cmd_resume(struct lldpctl_conn_t *conn, struct writer *w,
    struct cmd_env *env, void *arg) {
	(void)w; (void)env;
	return cmd_pause_resume(conn, 0);
}


#ifdef HAVE_LIBREADLINE
static int
_cmd_complete(int all)
{
	char **argv = NULL;
	int argc = 0;
	int rc = 1;
	size_t len = strlen(rl_line_buffer);
	char *line = malloc(len + 2);
	if (!line) return -1;
	strlcpy(line, rl_line_buffer, len + 2);
	line[rl_point]   = 2;	/* empty character, will force a word */
	line[rl_point+1] = 0;

	if (tokenize_line(line, &argc, &argv) != 0)
		goto end;

	char *compl = commands_complete(root, argc, (const char **)argv, all, is_privileged());
	if (compl && strlen(argv[argc-1]) < strlen(compl)) {
		if (rl_insert_text(compl + strlen(argv[argc-1])) < 0) {
			free(compl);
			goto end;
		}
		free(compl);
		rc = 0;
		goto end;
	}
	/* No completion or several completion available. */
	free(compl);
	fprintf(stderr, "\n");
	rl_forced_update_display();
	rc = 0;
end:
	free(line);
	tokenize_free(argc, argv);
	return rc;
}

static int
cmd_complete(int count, int ch)
{
	return _cmd_complete(0);
}

static int
cmd_help(int count, int ch)
{
	return _cmd_complete(1);
}
#else
static char*
readline(const char *p)
{
	static char line[2048];
	fprintf(stderr, "%s", p);
	fflush(stderr);
	if (fgets(line, sizeof(line) - 2, stdin) == NULL)
		return NULL;
	return strdup(line);
}
#endif

/**
 * Execute a tokenized command and display its output.
 *
 * @param conn The connection to lldpd.
 * @param fmt  Output format.
 * @param argc Number of arguments.
 * @param argv Array of arguments.
 * @return 0 if an error occurred, 1 otherwise
 */
static int
cmd_exec(lldpctl_conn_t *conn, const char *fmt, int argc, const char **argv)
{
	/* Init output formatter */
	struct writer *w;

	if      (strcmp(fmt, "plain")    == 0) w = txt_init(stdout);
	else if (strcmp(fmt, "keyvalue") == 0) w = kv_init(stdout);
	else if (strcmp(fmt, "json")     == 0) w = json_init(stdout, 1);
	else if (strcmp(fmt, "json0")    == 0) w = json_init(stdout, 0);
#ifdef USE_XML
	else if (strcmp(fmt, "xml")      == 0) w = xml_init(stdout);
#endif
	else {
		log_warnx("lldpctl", "unknown output format \"%s\"", fmt);
		w = txt_init(stdout);
	}

	/* Execute command */
	int rc = commands_execute(conn, w,
	    root, argc, argv, is_privileged());
	if (rc != 0) {
		log_info("lldpctl", "an error occurred while executing last command");
		w->finish(w);
		return 0;
	}
	w->finish(w);
	return 1;
}

/**
 * Execute a command line and display its output.
 *
 * @param conn The connection to lldpd.
 * @param fmt  Output format.
 * @param line Line to execute.
 * @return -1 if an error occurred, 0 if nothing was executed. 1 otherwise.
 */
static int
parse_and_exec(lldpctl_conn_t *conn, const char *fmt, const char *line)
{
	int cargc = 0; char **cargv = NULL;
	int n;
	log_debug("lldpctl", "tokenize command line");
	n = tokenize_line(line, &cargc, &cargv);
	switch (n) {
	case -1:
		log_warnx("lldpctl", "internal error while tokenizing");
		return -1;
	case 1:
		log_warnx("lldpctl", "unmatched quotes");
		return -1;
	}
	if (cargc != 0)
		n = cmd_exec(conn, fmt, cargc, (const char **)cargv);
	tokenize_free(cargc, cargv);
	return (cargc == 0)?0:
	    (n == 0)?-1:
	    1;
}

static struct cmd_node*
register_commands()
{
	root = commands_root();
	register_commands_show(root);
	register_commands_watch(root);
	commands_privileged(commands_new(
		commands_new(root, "update", "Update information and send LLDPU on all ports",
		    NULL, NULL, NULL),
		NEWLINE, "Update information and send LLDPU on all ports",
		NULL, cmd_update, NULL));
	register_commands_configure(root);
	commands_hidden(commands_new(root, "complete", "Get possible completions from a given command",
		NULL, cmd_store_env_and_pop, "complete"));
	commands_new(root, "help", "Get help on a possible command",
	    NULL, cmd_store_env_and_pop, "help");
	commands_new(
		commands_new(root, "pause", "Pause lldpd operations", NULL, NULL, NULL),
		NEWLINE, "Pause lldpd operations", NULL, cmd_pause, NULL);
	commands_new(
		commands_new(root, "resume", "Resume lldpd operations", NULL, NULL, NULL),
		NEWLINE, "Resume lldpd operations", NULL, cmd_resume, NULL);
	commands_new(
		commands_new(root, "exit", "Exit interpreter", NULL, NULL, NULL),
		NEWLINE, "Exit interpreter", NULL, cmd_exit, NULL);
	return root;
}

struct input {
	TAILQ_ENTRY(input) next;
	char *name;
};
TAILQ_HEAD(inputs, input);
static int
filter(const struct dirent *dir)
{
	if (strlen(dir->d_name) < 5) return 0;
	if (strcmp(dir->d_name + strlen(dir->d_name) - 5, ".conf")) return 0;
	return 1;
}

/**
 * Append a new input file/directory to the list of inputs.
 *
 * @param arg       Directory or file name to add.
 * @param inputs    List of inputs
 * @param acceptdir 1 if we accept a directory, 0 otherwise
 */
static void
input_append(const char *arg, struct inputs *inputs, int acceptdir, int warn)
{
	struct stat statbuf;
	if (stat(arg, &statbuf) == -1) {
		if (warn) {
			log_warn("lldpctl", "cannot find configuration file/directory %s",
			    arg);
		} else {
			log_debug("lldpctl", "cannot find configuration file/directory %s",
			    arg);
		}
		return;
	}

	if (!S_ISDIR(statbuf.st_mode)) {
		struct input *input = malloc(sizeof(struct input));
		if (!input) {
			log_warn("lldpctl", "not enough memory to process %s",
			    arg);
			return;
		}
		log_debug("lldpctl", "input: %s", arg);
		input->name = strdup(arg);
		TAILQ_INSERT_TAIL(inputs, input, next);
		return;
	}
	if (!acceptdir) {
		log_debug("lldpctl", "skip directory %s",
		    arg);
		return;
	}

	struct dirent **namelist = NULL;
	int n =	scandir(arg, &namelist, filter, alphasort);
	if (n < 0) {
		log_warnx("lldpctl", "unable to read directory %s",
		    arg);
		return;
	}
	for (int i=0; i < n; i++) {
		char *fullname;
		if (asprintf(&fullname, "%s/%s", arg, namelist[i]->d_name) != -1) {
			input_append(fullname, inputs, 0, 1);
			free(fullname);
		}
		free(namelist[i]);
	}
	free(namelist);
}

int
main(int argc, char *argv[])
{
	int ch, debug = 0, use_syslog = 0, rc = EXIT_FAILURE;
	const char *fmt = "plain";
	lldpctl_conn_t *conn = NULL;
	const char *options = is_lldpctl(argv[0])?"hdvf:u:":"hdsvf:c:C:u:";

	int gotinputs = 0, version = 0;
	struct inputs inputs;
	TAILQ_INIT(&inputs);

	ctlname = lldpctl_get_default_transport();

	signal(SIGHUP, SIG_IGN);

	/* Get and parse command line options */
	optind = 1;
	while ((ch = getopt(argc, argv, options)) != -1) {
		switch (ch) {
		case 'd':
			if (use_syslog)
				use_syslog = 0;
			else
				debug++;
			break;
		case 's':
			if (debug == 0)
				use_syslog = 1;
			else
				debug--;
			break;
		case 'h':
			usage();
			break;
		case 'u':
			ctlname = optarg;
			break;
		case 'v':
			version++;
			break;
		case 'f':
			fmt = optarg;
			break;
		case 'C':
		case 'c':
			if (!gotinputs) {
				log_init(use_syslog, debug, __progname);
				lldpctl_log_level(debug + 1);
				gotinputs = 1;
			}
			input_append(optarg, &inputs, 1, ch == 'c');
			break;
		default:
			usage();
		}
	}

	if (version) {
		version_display(stdout, "lldpcli", version > 1);
		exit(0);
	}

	if (!gotinputs) {
		log_init(use_syslog, debug, __progname);
		lldpctl_log_level(debug + 1);
	}

	/* Disable SIGPIPE */
	signal(SIGPIPE, SIG_IGN);

	/* Register commands */
	root = register_commands();

	/* Make a connection */
	log_debug("lldpctl", "connect to lldpd");
	conn = lldpctl_new_name(ctlname, NULL, NULL, NULL);
	if (conn == NULL) goto end;

	/* Process file inputs */
	while (gotinputs && !TAILQ_EMPTY(&inputs)) {
		/* coverity[use_after_free]
		   TAILQ_REMOVE does the right thing */
		struct input *first = TAILQ_FIRST(&inputs);
		log_debug("lldpctl", "process: %s", first->name);
		FILE *file = fopen(first->name, "r");
		if (file) {
			size_t n;
			ssize_t len;
			char *line;
			while (line = NULL, len = 0, (len = getline(&line, &n, file)) > 0) {
				if (line[len - 1] == '\n') {
					line[len - 1] = '\0';
					parse_and_exec(conn, fmt, line);
				}
				free(line);
			}
			free(line);
			fclose(file);
		} else {
			log_warn("lldpctl", "unable to open %s",
			    first->name);
		}
		TAILQ_REMOVE(&inputs, first, next);
		free(first->name);
		free(first);
	}

	/* Process additional arguments. First if we are lldpctl (interfaces) */
	if (is_lldpctl(NULL)) {
		char *line = NULL;
		for (int i = optind; i < argc; i++) {
			char *prev = line;
			if (asprintf(&line, "%s%s%s",
				prev?prev:"show neigh ports ", argv[i],
				(i == argc - 1)?" details":",") == -1) {
				log_warnx("lldpctl", "not enough memory to build list of interfaces");
				free(prev);
				goto end;
			}
			free(prev);
		}
		if (line == NULL && (line = strdup("show neigh details")) == NULL) {
			log_warnx("lldpctl", "not enough memory to build command line");
			goto end;
		}
		log_debug("lldpctl", "execute %s", line);
		if (parse_and_exec(conn, fmt, line) != -1)
			rc = EXIT_SUCCESS;
		free(line);
		goto end;
	}

	/* Then, if we are regular lldpcli (command line) */
	if (optind < argc) {
		const char **cargv;
		int cargc;
		cargv = &((const char **)argv)[optind];
		cargc = argc - optind;
		if (cmd_exec(conn, fmt, cargc, cargv) == 1)
			rc = EXIT_SUCCESS;
		goto end;
	}

	if (gotinputs) {
		rc = EXIT_SUCCESS;
		goto end;
	}

	/* Interactive session */
#ifdef HAVE_LIBREADLINE
	rl_bind_key('?',  cmd_help);
	rl_bind_key('\t', cmd_complete);
#endif
	char *line = NULL;
	do {
		if ((line = readline(prompt()))) {
			int n = parse_and_exec(conn, fmt, line);
			if (n != 0) {
#ifdef HAVE_READLINE_HISTORY
				add_history(line);
#endif
			}
			free(line);
		}
	} while (!must_exit && line != NULL);
	rc = EXIT_SUCCESS;

end:
	while (!TAILQ_EMPTY(&inputs)) {
		/* coverity[use_after_free]
		   TAILQ_REMOVE does the right thing */
		struct input *first = TAILQ_FIRST(&inputs);
		TAILQ_REMOVE(&inputs, first, next);
		free(first->name);
		free(first);
	}
	if (conn) lldpctl_release(conn);
	if (root) commands_free(root);
	return rc;
}
